File: Geometry/Vertex/alpha.cpp
#include "alpha.h"

#include <stdio.h>

typedef map<TriPosition, Alpha*, TriPositionCompare> AlphaIndex;
static AlphaIndex* Index = NULL;

Alpha::Alpha( Vertex& v ){}

void Alpha::recalculate(){
   // DEFAULT
   value = 1;
}

void Alpha::remove() {
     deleteDependents();
     Index->erase(pos);
     delete this;
}

Alpha::~Alpha(){}

Alpha* Alpha::At( Vertex& v ){
  TriPosition T( 1, v.getSerialNumber() );
  if( Index == NULL ) Index = new AlphaIndex();
  AlphaIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    Alpha* val = new Alpha( v );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void Alpha::CleanUp(){
  if( Index == NULL ) return;
  AlphaIndex::iterator iter;
  AlphaIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

void Alpha::print(FILE* out) {
  map<int, Vertex>::iterator vit;
  
  fprintf(out, "Alpha [ v ]\n=============\n");
  for(vit = Triangulation::vertexTable.begin(); vit != Triangulation::vertexTable.end(); vit++) {
     fprintf(out, "Alpha [%3d]\t= % 2.8f\n", vit->first, Alpha::valueAt(vit->second));
  }       
}

File: Java/src/Geoquant/Alpha.java
package Geoquant;

import java.util.HashMap;

import Triangulation.Vertex;

public class Alpha extends Geoquant {

  private static HashMap<TriPosition, Alpha> Index = new HashMap<TriPosition, Alpha>();
  
  private Alpha(Vertex v) {
    super(v); // ALWAYS have to call this first
    value = 0; // default
  }
  
  public static Alpha At(Vertex v) {
    TriPosition T = new TriPosition( v.getSerialNumber() );
    Alpha q = Index.get(T);
    if(q == null) {
      q = new Alpha(v);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Vertex v) {
    return At(v).getValue();
  }
  
  protected void recalculate() {
    // This is empty for a alpha
    value = 0;
  }
 
  public void remove() {
    deleteDependents();
    Index.remove(pos);
  }

}
